src/main/java/com/google/api/codegen/util/ProtoParser.java

```diff
@@ -422,6 +422,57 @@ private String getResourceSetFullName(ResourceSet resource, ProtoFile file) {
     return String.format("%s.%s", resource.getSymbol(), getProtoPackage(file));
   }
 
+  public ImmutableMap<String, String> getFieldNamePatterns(Method method) {
+    // Only look two levels deep in the request object, so fields of fields of the request object.
+    return ImmutableMap.copyOf(getFieldNamePatterns(method.getInputMessage(), 2));
+  }
+
+  private Map<String, String> getFieldNamePatterns(MessageType messageType, int depth) {
+    LinkedHashMap<String, String> builder = new LinkedHashMap<>();
+    for (Field field : messageType.getFields()) {
+      String fieldName = field.getSimpleName();
+
+      if (field.getType().isMessage() && depth > 0) {
+        Map<String, String> children =
+            getFieldNamePatterns(field.getType().getMessageType(), depth - 1);
+        for (Map.Entry<String, String> child : children.entrySet()) {
+          // Dot-separate the nested fields, e.g. "shelf.name"
+          builder.put(String.format("%s.%s", fieldName, child.getKey()), child.getValue());
```

Comment:
Suggestion - can we add a `builder` and a `prefix` argument to this method? That way we could avoid having to re-copy the children each time (adding the prefix) and also avoid the ImmutableMap.copyOf at the call site.

Reply:
yeah!